### PR TITLE
Allow rolling feature flags out to multiple groups

### PR DESCRIPTION
**Pull request by macobo**
## Changes

This makes it possible to roll out feature flags in several groups. #3007 

Since feature flags are used by /decide endpoint, just doing a data migration would mean dropped requests while new servers are spinning up => I added application-level logic for handling that. This logic can be removed/migrated at a later date.

## Commits

- Support multiple groups in feature flags model
- Update analytics_metadata method
- Allow defining multiple groups in frontend
- Allow sorting
- Update tests, allow submit without filters.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

File: posthog/api/feature_flag.py [hunk 1]
```diff
@@ -58,10 +55,15 @@ def update(self, instance: FeatureFlag, validated_data: Dict, *args: Any, **kwar
             validated_key = validated_data.get("key", None)
             if validated_key:
                 FeatureFlag.objects.filter(key=validated_key, team=instance.team, deleted=True).delete()
+            self._update_filters(validated_data)
```
Comment (EDsCODE) [inline posthog/api/feature_flag.py:58]: Does this need to be called twice?

**Review by EDsCODE — COMMENTED** [review]
A small UI consideration would be how these groups are represented in the table. The first feature flag is a 2 group feature flag but the rollout percentages aren't shown and the filters aren't clearly defined with an OR relationship. The second feature flag is just a single group with filters AND'd together

Screen Shot 2021-01-21 at 9 35 00 AM

File: posthog/api/feature_flag.py (hunk 1 above)
Comment (macobo) [reply posthog/api/feature_flag.py:58]: It's called once per method - once in create, once in update.

**Comment by macobo** [comment]
Good point re UX - not sure how to best represent it and release percentage within the table without getting too busy.

Suggestions/thoughts? :)

File: posthog/api/feature_flag.py (hunk 1 above)
Comment (EDsCODE) [reply posthog/api/feature_flag.py:58]: Oh wow that line break in dark mode is not clear on github review

**Comment by EDsCODE** [comment]
Two options come to mind:
1. low hanging fruit is just to  add the or/and tag in between filters and then iterate in order the percentages in the percentage column. This would cause some annoying cognitive load because you'd have to match up the group to the percentage
2. filters and rollout percentages should just be a single column called filters. Each filter is accompanied by a rollout percentage tag. Then you put the or/and between the filters

**Comment by macobo** [comment]
I tried. Getting all the filters displaying in the table just made the information density too high, went a different route:

**Comment by EDsCODE** [comment]
this is a lot better than the previous columns. What was immediately still confusing was that the percentage is shown in the first row when there's a single filter but it disappears when there are more filters such as in row 3 even though i'm assuming those are and'd together in the same group?

**Comment by macobo** [comment]
> this is a lot better than the previous columns. What was immediately still confusing was that the percentage is shown in the first row when there's a single filter but it disappears when there are more filters such as in row 3 even though i'm assuming those are and'd together in the same group?

Row 3 has no rollout percentage defined, which is why we don't show it.

**Review by EDsCODE — APPROVED** [review]
Got it. I'm good with this!

**Review by paolodamico — COMMENTED** [review]
Oops, seems like I forgot to submit my review yesterday. Anyways think these comments might be useful for a subsequent PR. 

Commenting only on the UI/UX, here are my 2c,
- This UI seems quite cluttered and hard to understand even on the best of circumstances.
    Screenshot 2021-01-25 at 6 25 42 PM
    I would suggest trying something like this, where we can still provide very useful information in the same page while making it clear what each group is comprised of. But this might entail updating all similar filtering related components,
    

- The add group button is hard to find, and it's not use is not very intuitive. In addition, it's not standardized with how we add match groups elsewhere in the app. Would suggest taking a look at actions or cohorts.
    

- I would definitely add the "OR" tag (as we do on actions) to make it extra clear how the match groups are referenced. Additionally the close button is also a bit hard to see (particularly the first times) and doesn't match the actions UI either. To be fair it does match the cohorts UI a bit more, but I think it might be worth leaning towards the actions approach.
    




Let me know if you want me to take on the UI improvements. I would need a couple of days to carve out some time for this but happy to do it.

**Pull request merged**